docs/proposals/ssh_adapter.md

```diff
@@ -0,0 +1,194 @@
+
+    $ ssh [{user}@]{server} git-lfs-transfer {path} {operation}
+
+If authentication fails, or some other connection error occurs, errors will be
+read fronm standard error and displayed to the user.  The operation may be
+`upload`, `download`, or `meta` (for items which are not transfers).
+
+Once the connection is established, the server should send a capability
+advertisement:
+
+```
+capability-advertisement = capability-list flush-pkt
+capability-list = *capability
+capability = PKT-LINE(key=[value] LF)
+
+key = 1*(ALPHA | DIGIT | "-_")
+value = 1*(ALPHA | DIGIT | " -_.,?\/{}[]()<>!@#$%^&*+=:;")
+```
+
+Unlike the Git protocol, but like IMAP, the protocol version is specified as a
+capability.  This document defines protocol version 1, which is specified as
+`version=1`.  If the server supports other protocol versions, it may enumerate
+them here as well.
+
+The client will then issue an appropriate version command:
+
+```
+version-request = PKT-LINE("version " number LF) flush-pkt
+number = 1*DIGIT
+```
+
+No reply is expected.
+
+### Requests to transfer objects
+
+These commands may be used if the operation was `upload` or `download`.
+
+The `batch` command is used to specify a JSON command identical those used at
+the `info/lfs/object/batch` endpoint:
+
+```
+batch-request = batch-command
+                *arguments
+                delim-pkt
+                json-message
+                flush-pkt
+batch-command = PKT-LINE("batch" LF)
+arguments = PKT-LINE(key=[value] LF)
+json-message = 1*PKT-LINE(json-data)
+```
+
+It is recommended to specify a `size` argument, which is the number of bytes in
+decimal of the JSON payload without pkt-line encoding.  A server may signal an
+error if it requires a `size` argument for an operation but one is not specified.
+
+Unknown arguments should be ignored.
+
+The response from the server will look like the following:
+
+```
+status-json-response = status-command
+                       *arguments
+                       delim-pkt
+                       json-message
+                       flush-pkt
+status-command = PKT-LINE("status " http-status-code LF)
+http-status-code = 3DIGIT
+```
+
+The `size` argument may be specified here as well.
+
+In addition to the already specified items in the `actions` hash in the JSON
+response, the response may include a string, `id`, which is an opaque identifier
+relevant only to the server to help it identify the object, and another string,
+`token`, which is an opaque identifier relevant only to the server to help it
+manage authentication.  These strings must meet the syntax for the `value`
+production above; if arbitrary bytes are needed, Base64 encoding is recommended.
+
+These values, if specified, must be passed as arguments to the `get-object` and
+`put-object` commands.
+
+The `http-status-code` portion of the response is an HTTP status code, identical
+to those used if the request is made over HTTP.
+
+### Downloads
```

Comment:
Ah, ignore my previous comments made about unifying the batch protocol into its individual components over the wire. If my understanding is correct, that is exactly what you aim to do here. In this case, what do you think about getting rid of the batch protocol entirely for the wire protocol?

I think that there is certainly an appealing quality about introducing something (even though we know that it will be deprecated eventually) that is designed to ease the burden of implementers. (At this point, I'm requesting :eyes: from @git-lfs/implementers.)

But, I'm worried that this is a deferred cost with high interest that allows us to put off the harder change, and makes the hard transition even harder. What do you think?

Reply:
As I mentioned above, I think the batch protocol is useful to determine whether we need to upload objects at all.  Since we'd like to avoid uploading unneeded objects, we should maintain the batch protocol or specify that the client has to verify objects before uploading them.  The latter potentially means a large number of round trips, which is bad for high-latency connections.  Maybe that's not very much of a concern for Git LFS, which kind of assumes that a stable, reasonably performant connection is available, but I can imagine people traveling on airplanes being unhappy about that.

I'm fine simply specifying that one should use `verify-object` if you think that's best.  If so, then I'll simply drop the `id` and `token` arguments as well.

Reply:
I think leaving it as-is is fine, though I'll defer to your judgement in the final case.